xfer/publisher.go

```diff
@@ -1,215 +1,10 @@
 package xfer
 
-import (
-	"bytes"
-	"fmt"
-	"log"
-	"net/http"
-	"net/url"
-	"strings"
-	"sync"
-	"time"
-)
+import "io"
 
-const (
-	initialBackoff = 1 * time.Second
-	maxBackoff     = 60 * time.Second
-)
-
-// Publisher is something which can send a buffered set of data somewhere,
-// probably to a collector.
+// Publisher is something which can send a stream of data somewhere, probably
+// to a remote collector.
 type Publisher interface {
-	Publish(*bytes.Buffer) error
+	Publish(io.Reader) error
```

Comment:
The primary reason is that a concrete bytes.Buffer is an overspecification for the interface. Fundamentally, all the Publisher is doing is reading bytes from the thing. Thus, io.Reader.

The "cost" is that we have to ioutil.ReadAll in the MultiPublisher, which invokes an extra copy() builtin compared to just bytes.Buffer.Bytes. But I judge it's a good tradeoff for a more-tightly-defined API.